pkg/asset/agent/agentconfig/agent_config.go

```diff
@@ -94,19 +100,78 @@ func (a *Asset) validateNodesHaveAtLeastOneMacAddressDefined() field.ErrorList {
 		return allErrs
 	}
 
+	rootPath := field.NewPath("Spec", "Hosts")
+
 	for i := range a.Config.Spec.Hosts {
 		node := a.Config.Spec.Hosts[i]
+		interfacePath := rootPath.Index(i).Child("Interfaces")
 		if len(node.Interfaces) == 0 {
-			interfacePath := field.NewPath("Spec", "Hosts", "Interfaces", "macAddress")
 			allErrs = append(allErrs, field.Required(interfacePath, "at least one interface must be defined for each node"))
 		}
 
 		for j := range node.Interfaces {
 			if node.Interfaces[j].MacAddress == "" {
-				macAddressPath := field.NewPath("Spec", "Hosts", "Interfaces", "macAddress")
+				macAddressPath := interfacePath.Index(j).Child("macAddress")
 				allErrs = append(allErrs, field.Required(macAddressPath, "each interface must have a MAC address defined"))
 			}
 		}
 	}
 	return allErrs
 }
+
+func (a *Asset) validateRootDeviceHints() field.ErrorList {
+	var allErrs field.ErrorList
+	rootPath := field.NewPath("Spec", "Hosts")
+
+	for i, host := range a.Config.Spec.Hosts {
+		hostPath := rootPath.Index(i)
+		if host.RootDeviceHints.WWNWithExtension != "" {
+			allErrs = append(allErrs, field.Forbidden(
+				hostPath.Child("RootDeviceHints", "WWNWithExtension"),
+				"WWN extensions are not supported in root device hints"))
+		}
+		if host.RootDeviceHints.WWNVendorExtension != "" {
+			allErrs = append(allErrs, field.Forbidden(
+				hostPath.Child("RootDeviceHints", "WWNVendorExtension"),
+				"WWN vendor extensions are not supported in root device hints"))
+		}
+	}
+
+	return allErrs
+}
+
+type hostConfigFileMap map[string][]byte
+
+// HostConfigFiles returns a map from filename to contents of the files used for
+// host-specific configuration by the agent installer client
+func (a *Asset) HostConfigFiles() (hostConfigFileMap, error) {
```

Comment:
I think that if an asset can produce additional entities (based on the ones natively managed) then this step should be placed in the `finish()` method, before the validation rules (and the entity exposed as fields). In this way the asset user will be ensured that when the dependency is retrieved all the validation rules will be applied, and in the positive case the fetched entities are good enough to be used without any additional check (for more complex cases the next refactoring step could be to create a new standalone asset for the additional entity)

Reply:
Everything _is_ validated when the asset is generated. Are you saying this function shouldn't return an `error`? I don't think it's possible for the yaml.Marshal to actually fail, given that it's marshalling data that was just unmarshalled from yaml.

Reply:
There's no way to generate an agent-config (other than an empty one), it's the start of the line.

I'm struggling to see any difference between getting the data from a field vs. a method, except that the former requires 2 copies of everything to be kept around in memory and written to the asset store. This method could just as easily have been on the Ignition asset, or the agent.Config type itself.

> Placing this method in the `finish()` (better _after_ the validation rules) wil ensure to generate an `HostConfigFileMap` and eventually fail, if required, regardless of the fact that the asset was loaded from disk or generated.

I don't understand this part. There's no way this method can fail, so who cares when it runs?

Reply:
My comment was general for a (writable) asset, and we agree that for this specifc case the `Generate` it's not used.
An asset field it's to ensure that its content has been properly validated during the fetch phase (and if there was an error the whole process was terminated). A non-getter method with an error in its signature requires the consumer to handle it (or the method shouldn't have exposed it).
The point about the asset store storage it's fair one, and would probably justify moving the `HostConfigFiles` method into the ignition asset, especially if it's the only consumer of such artifact (ie no other assets are requiring such map).